### Laboratorio 6 - Ejercicio 3
Stefano Aragoni, Carol Arévalo, Luis Santos, Diego Perdomo

-----------

`Ejercicio 3`

Tiene la tarea de modelar el crecimiento de la población en múltiples escalas, desde organismos individuales hasta la población global. Para esto debe implementar un modelo multiescala en Python para analizar el crecimiento poblacional y su impacto

-----------

##### **Modelación**

Modelo Multiescala Simple para el Crecimiento Poblacional:

Nivel individual: Cada individuo tiene una probabilidad de reproducción y una probabilidad de mortalidad basadas en su edad.

Nivel de comunidad local: La tasa de reproducción se ajusta según la densidad poblacional.

Nivel global: Se ajustan las tasas de crecimiento según las migraciones.

In [1]:
import random

class Individuo:
    def __init__(self, edad):
        self.edad = edad

    def reproducir(self):
        # Probabilidad de reproducción disminuye con la edad
        prob_reproduccion = max(0.2 - self.edad * 0.01, 0.05)
        return random.random() < prob_reproduccion

    def morir(self):
        # Probabilidad de mortalidad aumenta con la edad
        prob_mortalidad = self.edad * 0.02
        return random.random() < prob_mortalidad

    def envejecer(self):
        self.edad += 1

class Comunidad:
    def __init__(self, poblacion_inicial):
        self.individuos = [Individuo(random.randint(0, 50)) for _ in range(poblacion_inicial)]

    def paso_temporal(self):
        # Nivel individual
        nuevos_individuos = []
        for individuo in self.individuos:
            individuo.envejecer()
            if individuo.reproducir():
                nuevos_individuos.append(Individuo(0))
            if not individuo.morir():
                nuevos_individuos.append(individuo)
        self.individuos = nuevos_individuos

        # Nivel de comunidad local
        if len(self.individuos) > 1000:
            # Reducir la tasa de reproducción si la población es muy alta
            self.individuos = random.sample(self.individuos, int(len(self.individuos) * 0.9))

        # Nivel global (simplificado)
        # Supongamos que hay una migración neta de 10 individuos cada año
        migracion = 10
        self.individuos += [Individuo(random.randint(0, 50)) for _ in range(migracion)]

    def poblacion_total(self):
        return len(self.individuos)

# Simulación
comunidad = Comunidad(500)
for año in range(10):
    comunidad.paso_temporal()
    print(f"Año {año+1}: Población total = {comunidad.poblacion_total()}")


Año 1: Población total = 284
Año 2: Población total = 221
Año 3: Población total = 202
Año 4: Población total = 205
Año 5: Población total = 217
Año 6: Población total = 230
Año 7: Población total = 252
Año 8: Población total = 269
Año 9: Población total = 277
Año 10: Población total = 295


*Resumen y Análisis:*

Inicio de la Simulación: La población inicial era de 500 individuos.

Año 1: Se observa una disminución significativa en la población, llegando a 284 individuos. Esto puede ser el resultado de la mortalidad basada en la edad y la reducción de la tasa de reproducción debido a la alta densidad poblacional inicial.

Años 2 y 3: La población sigue disminuyendo, lo que indica que la tasa de mortalidad sigue siendo más alta que la tasa de reproducción, a pesar de la migración neta positiva.

Año 4: La población se estabiliza ligeramente, con un pequeño aumento. Esto puede ser el resultado de un equilibrio entre la tasa de reproducción y la tasa de mortalidad, junto con la migración.

Años 5 a 10: Se observa un crecimiento constante y gradual de la población. Esto sugiere que la comunidad ha alcanzado un equilibrio donde la tasa de reproducción, ajustada por la densidad poblacional y la migración, supera la tasa de mortalidad.

Tendencia General: Después de una disminución inicial en la población, la comunidad muestra signos de recuperación y crecimiento sostenido. Esto puede ser el resultado de la adaptación de la comunidad a las condiciones iniciales y la influencia de factores a nivel individual, local y global.

-----------

##### **Preguntas**



1. ¿Qué es el modelado multiescala y por qué es esencial para comprender sistemas complejos como el crecimiento de la población?

El modelado multiescala se refiere a la técnica de modelar un sistema en diferentes escalas, desde la más pequeña (micro) hasta la más grande (macro). Es una poderosa técnica computacional que nos permite simular y analizar sistemas complejos en diferentes niveles de detalle, desde la escala atómica y molecular, hasta el nivel macroscópico.

*obtenido de las presentación de Alberto



2. Describe el modelo a nivel individual en esta simulación multiescala. ¿Qué factores se consideran a escala individual?

El modelo a nivel individual en esta simulación multiescala se centra en el comportamiento y características de un único organismo. A esta escala, se consideran los siguientes factores:

- **Edad**: La edad del individuo juega un papel crucial en las probabilidades tanto de reproducción como de mortalidad.

- **Probabilidad de reproducción**: Esta probabilidad disminuye con la edad del individuo. Específicamente, la probabilidad inicial de reproducción se establece en 0.2 y disminuye en 0.01 por cada año de edad del individuo. Sin embargo, tiene un límite mínimo, y no disminuirá por debajo del 0.05, independientemente de la edad del individuo.

- **Probabilidad de mortalidad**: Esta probabilidad aumenta con la edad del individuo. La probabilidad de que un individuo muera en un año determinado es igual al 0.02 multiplicado por su edad. Esto significa que, a medida que un individuo envejece, es más probable que muera.

- **Envejecimiento**: Cada paso temporal (o "año" en la simulación) incrementa la edad del individuo en uno.

Estos factores trabajan juntos para determinar si un individuo se reproduce, si muere y cómo cambia su probabilidad de hacerlo con el tiempo. Al considerar estos factores a nivel individual, el modelo puede capturar variabilidad en la población y simular de manera más realista cómo diferentes individuos contribuyen al crecimiento o disminucion de la poblacion.

3. Discuta el modelo a escala de comunidad local. ¿Cómo incorpora interacciones entre individuos y factores ambientales?


En la simulación multiescala presentada, el modelo a escala de comunidad local se centra en cómo los individuos dentro de una comunidad interactúan entre sí y cómo la densidad poblacional afecta la dinámica de la comunidad. A continuación, se describen las características y factores considerados a esta escala:

- **Densidad poblacional y capacidad de carga:** Si la población dentro de la comunidad supera un cierto umbral (en este caso, 1000 individuos), se realiza un ajuste para reducir la población. Específicamente, se toma una muestra aleatoria del 90% de la población actual, reduciendo así la población en un 10%. Este mecanismo puede simular la idea de capacidad de carga (un límite en el número de individuos que un ambiente puede soportar sin que se degrade). 

- **Interacción entre individuos:** Aunque no se modelan explícitamente interacciones individuales complejas, la dinámica de reproducción y mortalidad de cada individuo influye en la composición general y la estructura de edad de la comunidad. Por ejemplo, si muchos individuos en la comunidad son viejos y, por lo tanto, tienen una alta probabilidad de mortalidad y una baja probabilidad de reproducción, esto influirá en el crecimiento neto de la comunidad.

4. Explique el modelo a escala global. ¿Qué parámetros y factores se consideran al modelar toda la población mundial?

En la simulación multiescala presentada, el modelo a escala global se simplifica bastante y se centra principalmente en la migración:

- **Migración:** En el modelo, se asume una migración neta positiva hacia la comunidad, lo que significa que más individuos entran a la comunidad de los que salen. Esta migración se modela como un aumento fijo de 10 individuos en la comunidad cada año. Cada individuo que migra tiene una edad asignada aleatoriamente entre 0 y 50 años.

A pesar de su simplicidad, este componente del modelo reconoce la importancia de los movimientos poblacionales de una comunidad. La migración puede influir en la demografía y la dinámica de la población de una comunidad, y puede ser causada por una variedad de factores, como búsqueda de recursos, cambios climáticos o eventos socioeconómicos.

Sin embargo, es importante notar que, en un modelo más detallado del crecimiento poblacional a escala global, se considerarían muchos otros factores, como:

- **Emigración:** Al igual que algunos individuos pueden migrar hacia una comunidad, otros pueden salir de ella. Esto podría modelarse con una tasa o número fijo, o podría variar según las condiciones de la comunidad.
  
- **Eventos globales:** Factores como desastres naturales, pandemias, guerras o cambios económicos y políticos pueden tener un impacto significativo en las tasas de migración y en la dinámica poblacional general.
  
- **Interacciones entre comunidades:** Las comunidades interactúan entre sí, y estas interacciones pueden influir en la dinámica de cada comunidad individual.

- **Recursos y capacidad de carga global:** La disponibilidad global de recursos (como alimentos, agua, refugio) y la capacidad de carga del planeta o región también pueden afectar la dinámica poblacional.